compiler/rustc_monomorphize/src/collector/abi_check.rs

```diff
@@ -0,0 +1,160 @@
+            ABI_UNSUPPORTED_VECTOR_TYPES,
+            CRATE_HIR_ID,
+            span,
+            AbiErrorDisabledVectorTypeDef { span, required_feature },
+        );
+    })
+}
+
+fn call_site_abi_missing_features<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    (callee, caller): (Ty<'tcx>, DefId),
+) -> Vec<String> {
+    let mut answer = vec![];
+    let param_env = ParamEnv::reveal_all();
+    let callee_abi = match *callee.kind() {
+        ty::FnPtr(..) => {
+            tcx.fn_abi_of_fn_ptr(param_env.and((callee.fn_sig(tcx), ty::List::empty())))
+        }
+        ty::FnDef(def_id, args) => {
+            // Intrinsics are handled separately by the compiler.
+            if tcx.intrinsic(def_id).is_some() {
+                return vec![];
+            }
+            let instance = ty::Instance::expect_resolve(tcx, param_env, def_id, args, DUMMY_SP);
+            tcx.fn_abi_of_instance(param_env.and((instance, ty::List::empty())))
+        }
+        _ => {
+            panic!("Invalid function call");
+        }
+    };
+
+    let Ok(callee_abi) = callee_abi else {
+        // ABI failed to compute; this will not get through codegen.
+        return vec![];
+    };
+    do_check_abi(tcx, callee_abi, caller, |required_feature| {
+        answer.push(required_feature.to_string());
+    });
+    answer
+}
+
+/// Checks that a call expression does not try to pass a vector-passed argument which requires a
+/// target feature that the caller does not have, as doing so causes UB because of ABI mismatch.
+pub(super) fn check_call_site_abi<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    callee: Ty<'tcx>,
+    span: Span,
+    caller: InstanceKind<'tcx>,
+) {
+    if callee.fn_sig(tcx).abi().is_rust() {
+        // "Rust" ABI never passes arguments in vector registers.
+        return;
+    }
+    for required_feature in tcx.call_site_abi_missing_features((callee, caller.def_id())) {
+        tcx.emit_node_span_lint(
+            ABI_UNSUPPORTED_VECTOR_TYPES,
+            CRATE_HIR_ID,
+            span,
+            AbiErrorDisabledVectorTypeCall { span, required_feature },
+        );
+    }
+}
+
+pub(super) fn provide(providers: &mut Providers) {
+    *providers = Providers { check_instance_abi, call_site_abi_missing_features, ..*providers }
+}
+
+pub(super) fn should_check_instance_abi<'tcx>(tcx: TyCtxt<'tcx>, instance: Instance<'tcx>) -> bool {
+    // We do not need to check the instance for feature-dependent ABI if we can determine that the
+    // function has "Rust" ABI, which is known not to be feature-dependent.
+    // Note that the check is still correct on Rust ABI functions, but somewhat expensive. Hence,
+    // checking for "Rust" ABI is just an optimization.
+    // We also avoid to try to determine the type of the instance, as doing so involves running a
+    // query that does not usually run for unchanged functions in incremental builds.
+    if let InstanceKind::Item(def) | InstanceKind::ReifyShim(def, ..) = instance.def {
```

Comment:
Do you mind making this an exhaustive `match`?

Reply:
Done!

And while doing that, I also realized that there's a couple more cases in which I can omit the check, so perf might have improved slightly...